README
{{extension
|name=BugzillaReports
|author=[[User:Ian Homer|Ian Homer]]
|type=parser function
|version=1.1-beta1
|update=June 14th, 2009
|status=stable
|mediawiki=1.13
|license=GPLv2
|description=Generate bugzilla reports
|download=[http://blog.bemoko.com/bugzillareports/ BugzillaReports]
}}
'''BugzillaReports''' generates great looking reports from [[Bugzilla]] which you can include in your [[MediaWiki]] pages (or in fact any PHP application). Great to provide access to Bugzilla info, for example:

# Generate release notes when you release
# Generate roadmap reports for what's coming in a future release
# Generate your personal task radar - what should you be doing next
# Access controlled reports - if you want to expose bugzilla summaries to those who don't have access to Bugzilla
# Provide reports easy for non-techies to read. 

Displays dependency reports, last comments on bug on hover over, supports interactions with voting & flagging and helps you to see what needs to be done next.  Read below for more bells and whistles.  Makes reporting on your bugs much easier than through the standard Bugzilla search UI.  It's your controlled window onto Bugzilla ...

[[Category:Bugzilla]]

See [http://blog.bemoko.com/2008/06/28/bugzilla-reports-for-mediawiki here] for an active discussion thread on this extension.

''I expect to release [[#Roadmap|v0.9.8]] within the next week.  If you have requests for what will be included in the release after this one then email me at bugzillareports -at- bemoko.com if you want to add anything to this list (or ... find any bugs).  I'm also preparing the v1.0 (stable) release.  For this v1.0 release I'd welcome any corrections updates to this page to get these docs up to date - lots of features have gone in the past few months that documented in the release notes at the bottom of this page, but not in the "learn by examples" below.  Anyone who does some updates to the docs will get in the credits.''  --[[User:Ian.homer|Ian Homer]] 22:20, 29 November 2008 (UTC)

=Pre-requisites=

* Bugzilla 3.0.x
* MediaWiki 1.13
* MySQL or PostgreSQL

=Installation=

# Download the extension bundle from [http://blog.bemoko.com/bugzillareports/ BugzillaReports]
# Unpack bundle into your '''extensions''' directory.  This should give you
#: BugzillaReports/BugzillaReports.i18n.php
#: BugzillaReports/BugzillaReports.php
#: BugzillaReports/skins/bz_main.css
#: etc

Add the following to your LocalSettings.php file:

<source lang="php">
require_once("$IP/extensions/BugzillaReports/BugzillaReports.php");
$wgBugzillaReports = array(
  'host'        => "localhost", 
  'database'    => "bugzilla",
  'user'        => "user",
  'password'    => "(your password)",
  'bzserver'    => "http://host"
);
</source>

or if you want to use [[interwiki]] links (which is recommended and worth the extra effort of setting up)

<source lang="php">
...
  'interwiki'   => "bugzilla"
...
</source>

where parameters are defined as follows:

# '''host''' &rArr; defaults to localhost, but if Bugzilla not on same machine as your mediawiki install then set the hostname here
# '''user''' and '''password''' of the database user with read access to the bugzilla '''database'''
# To get the bugs to link through to Bugzilla server, use one of:
## '''bzserver''' &rArr; the URL of the bugzilla server for which the click through links should point to e.g. http://host".  Note that the path "/show_bug.cgi?id=" is automatically added for you by the extension
## '''interwiki''' &rArr; [[interwiki]] prefix linking to bugzilla instance
##: For example set up interwiki prefix of '''bugzilla''' to point to URL http://host/show_bug.cgi?id=$1
# '''maxrows''' &rArr; set maximum number of rows to display (defaults to 200)

==PostgreSQL==

<source lang="php">
$wgBugzillaReports = array(
  'host'      => "localhost" 
  'dbdriver'  => "pg",
  'database'  => "bugzilla",
  'user'      => "bugs",
  'password'  => "password",
  ...
);
</source>

=Example Output=

[[Image:bugzilla-reports-screenshot3.png|Bugzilla Report]]

Note that the bubble appears as you roll over a bug that has a last comment set.

=Usage=

: ''Note that several features are described in the release notes below and have not yet been documented in this section.  I anticipate reflecting these changes in the documentation prior to the v1 release.  Apologies for this gap in information, it should only be temporary'' --[[User:Ian.homer|Ian Homer]] 22:10, 21 October 2008 (UTC)

==By Example==
Generate a report from Bugzilla based on search criteria, e.g.

[[Image:bugzilla-reports-screenshot1.png|Bugzilla Report|right]]

List all open bugs - note that status defaults to NEW,ASSIGNED,UNCONFIRMED,REOPENED and maximum of 200 rows (by default) are returned. Results are sorted by priority. 

<source lang="php">
{{#bugzilla:}}
</source>

Report on P3 and higher bugs

<source lang="php">
{{#bugzilla:
  |priority=P1,P2,P3
}}
</source>

Report on a particular product version

<source lang="php">
{{#bugzilla:
  |product=myproduct
  |version=1.3
  |lastcomment=1
}}
</source>

Report on specific bug ids
<source lang="php">
{{#bugzilla:id=30,56,78,93}}
</source>

[[Image:bugzilla-reports-screenshot2.png|Last Comment Bubbles|right]]
Report on all enhancements with the text wiki in the summary with last comment bubbles

<source lang="php">
{{#bugzilla:
  |search=wiki
  |severity=enhancement
  |header=hide
  |lastcomment=1
}}
</source>

Setting the '''lastcomment''' parameter will generate a report which renders the last comment in bundles in bubbles as you hover each row (see image left)  This has been tested on Firefox 3 and Safari 3, but this option may have issues with other and older browser.  I'd be happy to hear feedback on this feature.

Report on all bugs that are not new

<source lang="php">
{{#bugzilla:status=!NEW}}
</source>

Everything except P1 and P2 bugs

<source lang="php">
{{#bugzilla:
  |priority=!(P1,P2)
  |status=*
}}
</source>

Set the message when no results are returned

<source lang="php">
{{#bugzilla:
  |product=doesntexist
  |noresultsmessage=all tasks complete
}}
</source>

Tasks with votes against them

<source lang="php">
{{#bugzilla:columns=+votes|votes=+}}
</source>

[[Image:bugzilla-reports-screenshot4.png]]

<source lang="php">
{{#bugzilla:
  |group=deadline
  |groupformat=radar
}}
</source>

[[Image:bugzilla-reports-screenshot5.png]]

<source lang="php">
{{#bugzilla:
  |group=version
  |bar=status
}}
</source>

All tasks created by me but not assigned to me

<source lang="php">
{{#bugzilla:columns=+to|from=me@bemoko.com|to=!me@bemoko.com}}
</source>

All tasks I'm cc'd on but not created by me and not assigned to me

<source lang="php">
{{#bugzilla:columns=+to,+from
  |to=!me@bemoko.com
  |from=!me@bemoko.com
  |cc=me@bemoko.com}}
</source>

Display list in a format that's handy for cutting and pasting into emails

<source lang="php">
{{#bugzilla:format=list}}
</source>

for example:

: [IHO] (2008-07-22) Do something (#19)
: [IHO] (2008-07-25) Do something else (#45)
: [JBL] Don't forget this (#54)

===Task Radar===

<source lang="php">
{{task radar|me@bemoko.com}}
</source>

where task radar is defined as:

<source lang="php">
=Flagged=
{{#bugzilla:flag={{{1}}}|lastcomment=1|noresultsmessage=No tasks flagged}}
=Up & Coming=
{{#bugzilla:to={{{1}}}|columns=-version,+deadline|deadline=+|group=deadline|groupformat=radar|lastcomment=1|noresultsmessage=Nothing up & coming}}
=My Product Priority Radar=
{{#bugzilla:to={{{1}}}|lastcomment=1|product=myproduct|noresultsmessage=Nothing to do on myproduct}}
=Voted Tasks=
{{#bugzilla:to={{{1}}}|columns=+deadline|group=product|votes=+|noresultsmessage=None of your tasks have received votes}}
=Raised=
{{#bugzilla:from={{{1}}}|to=!{{{1}}}|noresultsmessage=You have not raised anything}}
=CC=
{{#bugzilla:to=!{{{1}}}|from=!{{{1}}}|cc={{{1}}}|noresultsmessage=No ccs}}
=Other Priority Radar=
{{#bugzilla:to={{{1}}}|product=!myproduct|group=product|noresultsmessage=Nothing else to do}}
</source>

==Parameter Usage==

{| style="-moz-border-radius:8px;border: 1px solid #f0f0f0;border-spacing:0px;width: 80%;margin:auto"
|- style="vertical-align:top; background-color: #c0c0c0;"
! parameter
! style="width:30%" |values
! description
|- style="vertical-align:top;"
| '''alias'''
| alias value(s)
| alias query field
|- style="vertical-align:top; background-color: #f0f0f0;"
| '''bar'''
| column name
| field to summarise in a bar chart - note that this also works in conjunction with the group parameter to provide a set of bar charts.
|- style="vertical-align:top;"
| '''blocks'''
| id of blocking task (or "*" to include blocking tasks inline)
| blocks query field
|- style="vertical-align:top; background-color: #f0f0f0;"
| '''cc'''
|  single username
| Query for any task cc'd to the specified username, only supports single value
|- style="vertical-align:top;"
| '''columns'''
|  id, cc, deadline, modified, priority, product, severity, status, summary, to, version, estimated, remaining
| comma separated list of columns to display in the specified order
|- style="vertical-align:top; background-color: #f0f0f0;"
| '''component'''
| component name
| component query field
|- style="vertical-align:top;"
| '''deadline'''
| "*" &rArr; (any), "+" &rArr; (set)
| deadline query field
|- style="vertical-align:top;background-color: #f0f0f0;"
| '''debug'''
| true
| Set to enable debugging - outputs SQL at end of table
|- style="vertical-align:top;"
| '''depends'''
| id of dependent task (or "*" to include dependent tasks inline)
| dependent query field
|- style="vertical-align:top;"
| '''dependsstatus'''
| status query field
| Specify criteria for display depends (and blocks rows), e.g. dependsstatus=!(CLOSED,VERIFIED) will not display CLOSED and VERIFIED tickets in dependency or blocks report.  It defaults to !(CLOSED,VERIFIED,RESOLVED)
|- style="vertical-align:top; background-color: #f0f0f0;"
| '''detailsrow'''
| ''same as values for columns''

RELEASE-NOTES
=Release Notes=

==v1.3 - 4th August 2013==

* Minor fix for compatibility with MediaWiki 1.21
* Fix for compatibility with Bugzilla 4.x. Will no longer work with 3.x.

==v1.2==

* BUG : [#26] - Bugzilla search link not provided if no parameters are applicable to give a search
* BUG : [#27] - Fixed bug that led to some columns that were marked for removal weren't removed
* BUG : [#33] - Assigned and closed fields not working, throwing SQL exception
* BUG : [#39] - Add support for explicit style on report table, e.g {{#bugzilla:style=width:100%}}
* BUG : [#49] - Update how parser functions are registered to latest MediaWiki techniques

==v1.1 - 7th July 2009==

* [#1]  - Add safety check that you're not on a really old release of MediaWiki
* [#4]  - Add support for inline of totals, e.g. E,W,R = {{#bugzilla:columns=+estimated,+work,+remaining|total=estimated,work,remaining|hide=true|headers=hide|format=inline}}
* [#4]  - Totals are reported as zero if no rows found (as opposed to blank, with no tickets found message)
* [#8]  - Add a format to inline a bug report, e.g. {{#bugzilla:debug=1|id=661|status=*|format=inline}}
* [#13] - Added support for created filters, e.g. {{#bugzilla:created=-1m|createdformat=date}}
* [#13] - Added support for filters on status change, e.g. {{#bugzilla:assigned=2009-05-01:2009-05-31}}, 
  {{#bugzilla:resolved=2009-05-01:2009-05-31}}, {{#bugzilla:verified=2009-05-01:2009-05-31}},
  {{#bugzilla:closed=2009-05-01:2009-05-31}}, {{#bugzilla:reopened=2009-05-01:2009-05-31}}
* [#18] - Add support for attachments, e.g. {{#bugzilla:sort=attachments|order=desc|status=*}}
* [#21] - Added support for reports on hardware and os, e.g.   {{#bugzilla:columns=+hardware,os}}
  {{#bugzilla:columns=+hardware,os|os=Linux}}, {{#bugzilla:columns=+hardware,os|hardware=Macintosh}}
* [#25] - Sortable set to true by default, if you want to disable the sortable links then set "sortable=0". Sorting now works for all columns even if details row is displayed, but sorting disabled when grouping is enabled
* BUG : [#3]  - Task repeated in report for tasks with multiple keywords when keywords displayed in report
* BUG : [#5]  - deadline (or any date) range matches did not work, e.g. deadline=2008-12-01:2008-12-31
* BUG : [#6]  - bugzilla query link not generated correctly for id matches, e.g. {{#bugzilla:status=*|id=1,2,3}}
* BUG : [#11] - Support html in comments for last comment rendering, e.g. <table>
* BUG : [#12] - "worked_hours" field will be multiplied by the number of keywords you have, e.g. {{#bugzilla:debug=1|columns=+work,+keywords|status=*|keywords=analytics|sort=id|total=work}}
* BUG : [#17] - Support lastcomment in bugzilla 2.22 as well as 3.x
* BUG : [#19] - bzurl wrong when using "blocks" criteria
* BUG : [#20] - Fails to compute trigraph when name contains only one word 
* BUG : [#22] - support wiki style equals in comments without affecting the page TOC
* BUG : [#24] - support "-" in database names by enclosing database name in backticks in generated SQL

==v1.0 - 12th February 2009==

* Style class for severity set for the the tr element (row element) containing the bug details, so that (if you want) you can style based on severity.  Style classes used are the exactly the same name as the severity value, e.g. "enhancement", "normal", "major", etc
* Support for parameter "zeroasblank=false" to display numbers as "0" if it is actually 0.  The default setting is to hide 0s
* You can now set default parameters in the $wgBugzillaReports array defined in the LocalSettings.php for the whole wiki.  This is great if you don't like the defaults out-of-the-box and you want to override, e.g. 'default'       => "debug=1|zeroasblank=false|lastcomment=1".  Note that if the parameter is set in the function call (i.e. on the wiki page) then this will override the default setting in the LocalSettings.php
* Slight performance increase by switching to explode instead of split where appropriate
* Support two modes for handling of custom field naming; (1) implicitcustom=false (default) - custom fields must start with a given prefix, as defined by customprefix (which defaults to "cf_") or defined in the comma separated list of defined in the parameter customfields - note that nowadays Bugzilla enforces that custom fields start with "cf_".  E.g "columns=+cf_mycustom:MyCustom" (2) implicitcustom=true - if a column/field is referenced that is not in the supported parameter list then it is assumed to a custom field.  The supported parameter list is the list of all parameters that this extension accepts - note that their is a risk of future conflicts if you use a custom field name of parameter that is later used as a parameter for this extension.  e.g. "columns=+mycustom:MyCustom|implicitcustom=true".   This setting can be applied in your global LocalSettings file by including "implicitcustom=true" in the in the "default" parameter of the $wgBugzillaReports array.  This will set this mode across your whole wiki.

BugzillaQueryRenderer.php
<?php
/**
 * Bugzilla Query Renderer
 */

/**
 * Copyright (C) 2008 - Ian Homer & bemoko
 *
 * This program is free software; you can redistribute it and/or modify it 
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) 
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for    
 * more details.
 * 
 * You should have received a copy of the GNU General Public License along 
 * with this program; if not, see <http://www.gnu.org/licenses>.
 */

class BugzillaQueryRenderer {
  var $query;
  var $context;
  var $connector;
  var $output;
  
  public function BugzillaQueryRenderer($query) {
    $this->query=$query;
    $this->context=$query->context;
    $this->connector=$query->connector;
  }
  
  /**
   * Render query record set
   *
   * @param recordset
   * @return rendered markup
   */
  public function renderHTML($result) {
    $this->output = "";
    $nRows=$this->connector->getRowCount($result);
    
    $this->context->debug &&
      $this->context
        ->debug("Rendering report : rows = $nRows");

    if ($this->query->get('heading')) {
      $this->output.="<h1>".$this->query->get('heading')."</h1>";
    }
    # Table start
    if ($this->query->get('format') != "inline") {
        $this->output .= "<table class=\"bugzilla";
      if ($this->query->is('sortable')) {
        $this->output.=" sortable";
      }
      if ($this->query->get('bar')) {
        $this->output.=" bz_bar";
      }
      $this->output.="\"";
      if ($this->query->get('style')) {
        $this->output.=" style=\"".$this->query->get('style')."\"";
      }
      $this->output.=">";
    } else {
      
    }
    # Initialise details row logic
    $detailsRowColumns=array();
    $arrayOfDetailRowColumns=explode(",",$this->query->get('detailsrow'));
    foreach ($arrayOfDetailRowColumns as $detailRowColumn) {
      $detailsRowColumns[$detailRowColumn]=1;
    }
    $this->numberOfMainRowColumns=0;
    if ($this->query->get('bar')) {
      $this->numberOfMainRowColumns=2;
    } else {
      foreach ($this->query->getColumns() as $column) {
        if (!array_key_exists($column,$detailsRowColumns)) {
          $this->numberOfMainRowColumns++;
        }
      }
    }
    
    # Initialise total logic
    $arrayOfTotalColumns;
    $arrayOfTotals;
    if ($this->query->get('total')) {
      $arrayOfTotalColumns=explode(",",$this->query->get('total')); 
      $arrayOfTotals=array();   
      foreach ($arrayOfTotalColumns as $totalColumn) {
        $arrayOfTotals[$totalColumn]=0;
      }
    }
      
    # Display table header
    if ($this->query->get('headers')=="show" || (!$this->query->get('bar') && 
        $this->query->get('format')=='table' &&
        $this->query->get('headers')!="hide" &&
        !$this->query->explitlyOneValue)) {
        $this->output .= "<tr>";
      foreach ($this->query->getColumns() as $column) {
        $name=trim($column);
        if (array_key_exists($column,$this->query->columnName)) {
          $name=$this->query->columnName[$column];
        }
        $title;
        
        if (array_key_exists($column,$this->query->columnLabelName)) {
          $title=$this->query->columnLabelName[$column];
        } else {
          $title=$name;
        }
        $this->context->debug && 
          $this->context->debug("Name for column $column is $name");  
        if (($column=="summary") && ($nRows > 1)) {
          $name.=" ($nRows tasks)";
          if ($this->context->bzserver && $this->query->get('bzurl')=="show"
              && substr($this->query->bzURL,strlen($this->query->bzURL) -
              strlen(BugzillaQuery::$buglistServerRelativeUri) ) !=
              BugzillaQuery::$buglistServerRelativeUri) {
            if ($this->context->rawHTML) {
              $name.=" <a href=\"".$this->query->bzURL."\">&rArr;</a>";
            } else {
              $name.=" [".$this->query->bzURL." &rArr;]";
            }
          }
        }     
            $this->output .= "<th title=\"$title\"><b>$name</b></th>";
        }
        $this->output .= "</tr>";
    }

      # Create Table Data Rows
    $even=true;
    $count=0;
    $localMaxRows=$this->query->getMaxRows();
    $localMaxRowsForBarChart=$this->query->getMaxRowsForBarChart();
    $groupValue=array();
    $groupTotal=array();
    $groups;
    $doGrouping=0;
    if ($this->query->getGroup()) {
      $doGrouping=1;
      $groups=split(",",$this->query->getGroup());
      #
      # Prepare group counters
      #
      foreach ($groups as $group) {
        array_push($groupValue,"");
        array_push($groupTotal,0);
      }
    }
    $barArray=$this->getBarArray();
    $currentId=0;
    $currentBlocksId=0;
    $currentDependsId=0;
    while ($line = $this->connector->fetch($result)) {
      #
      # Add group heading
      #
      if ($doGrouping) {
        $iGroup=0;
        foreach ($groups as $group) {         
          $groupTotal[$iGroup]++;
          $newGroupValue=$this->query->formatForHeading(
            $line[$this->query->mapField(
              $group)],$group);
          if ($newGroupValue != $groupValue[$iGroup]) {
            $colspan=$this->numberOfMainRowColumns;
            if ($this->query->get('bar')) {
              if ($groupValue[$iGroup]) {
                $this->renderBar($barArray);
                foreach (array_keys($barArray) as $key) {
                  $barArray[$key]=0;
                }
              }
              $colspan-=1;
            }
            if ($this->query->get('total') && $groupValue) {
              $this->renderTotal($arrayOfTotals,1);         
              foreach ($arrayOfTotalColumns as $totalColumn) {
                $arrayOfTotals[$totalColumn]=0;
              }
            }
            $groupValue[$iGroup]=$newGroupValue;
            $this->output.="<tr class=\"bz_group bz_group_$iGroup\"><td colspan=\""
              .$colspan.
              "\">".$groupValue[$iGroup].
              "</td>";
            if ($this->query->get('bar')) {
              $this->output.="<td class=\"total\">&nbsp;</td>";
            }
            $this->output.="</tr>";
            $groupTotal[$iGroup]=0;
            break;
          }
          $iGroup++;
        }
      }
      ++$count;

      # Safety check break out of loop if there's too much.  Only
      # display warning if max rows has not come the function call, i.e.
      # user has explicitly truncated
      if ($count > $localMaxRows) {
        if ($localMaxRows > $this->query->get('maxrows')) {
          $this->context->
            warn("Report truncated - count greater than ".
              "max allowed ".
              "$count > $localMaxRows");
        }
        break;
      }
      #
      # Only render the row if the ID has changed from previous row
      # to support LEFT JOINS
      #
    if ($line["id"] != $currentId) {
      $this->context->debug &&
        $this->context
          ->debug("Rendering rows $count/$nRows : id =".
            $line["id"]);

      # Bar counter
      if ($this->query->get('bar')) {
        if ($count > $localMaxRowsForBarChart) {
          $this->context->warn("Bar count greater than max allowed ".
            "$count > $localMaxRowsForBarChart");
          break;
        }
        $barValue=$line[$this->query->mapField($this->query->get('bar'))];
        if (array_key_exists($barValue,$barArray)) {
          $barArray[$barValue]+=1;
        } else {
          $barArray[$barValue]=1;
        }
        continue;
      }
      
      # Total counter
      if ($this->query->get('total')) {
        foreach ($arrayOfTotalColumns as $totalColumn) {
          $value=$line[$this->query->mapField($totalColumn)];
          if (is_numeric($value)) {
            $arrayOfTotals[$totalColumn]+=$value;
          } else {
            $arrayOfTotals[$totalColumn]+=1;
          }
        }       
      }

      $currentId=$line["id"];
      $firstcolumn=true;
      $even=!$even;
      $class="bz_bug ".$line["priority"]." ".$line["severity"]." ".$line["status"]." ";
      if ($even) {
        $class.="bz_row_even";
      } else {
        $class.="bz_row_odd";
      }
      #
      # Formatting for inline and list
      #
      if (!$this->query->get('hide')) {
        if ($this->query->get('format') == "list" || $this->query->get('format') == "inline") {
          if ($this->query->get('format') == "list") {
            $this->output .= "<tr class=\"".$class."\">";         
            $this->output.="<td class=\"bz_list\" colspan=\"".
              $this->numberOfMainRowColumns."\">";
          } else {
            $this->output .= "<span>";
            if ($count > 1) {
              $this->output .= ", ";            
            }
          }
          if (array_key_exists("to",$this->query->getColumns())) {    
            $this->output.=
              "[".$this->query->
                format($line[$this->query->mapField('to')],
                'to',"")."] ";
          }
          if (array_key_exists("deadline",$this->query->getColumns())) {
            $deadline=$line[$this->query->mapField('deadline')];
            if ($deadline) {
              $this->output.=
                "(".$this->query->format($deadline,"deadline","").
                ") ";
            }
          }
          $this->output.=           
            $this->query->
              format($this->query->getDBValue($line,
                $this->query->mapField('summary')),
              'summary',"")." (#".
            $this->query->
              format($line[$this->query->mapField('id')],
              'id',"").")";
          if ($this->query->get('format') == "list") {
            $this->output .= "</td></tr>";
          } else {
            $this->output.="</span>";
          }
        #
        # Default formatting
        #
        } else {
            $this->output .= "<tr class=\"".$class."\">";
          foreach ($this->query->getColumns() as $column) {
            $dbColumn=$this->query->mapField($column);
            if (!$this->query->get('detailsrow') 
                or !array_key_exists($dbColumn,
                  $detailsRowColumns)) {
              $title=$this->query->getValueTitle($line,$dbColumn);
              $value=$this->query->
                format($this->query->getDBValue($line,$dbColumn),$column,$title);
              $this->output.="<td";
              if ($title) {
                $this->output.=" title=\"$title\"";
              }
              $this->output.=">$value";
              if ($firstcolumn) {
                $firstcolumn=false;
                /**
                 * Start with a carriage return so that comments starting with 
                 * list characters, e.g. *, # render in wiki style
                 */
                if ($this->query->get('lastcomment') or $this->query->get('keywords')) {
                  $lastcomment="";
                  $keywords="";
                  if ($this->query->get('lastcomment')) {
                    $lastcomment=trim($this->query->getDBValue($line,"thetext"));                   
                  }
                  if ($this->query->isRequired("keywords")) {
                    $keywords=$this->query->
                      format($line[$this->query->mapField("keywords")],
                        "keywords","");
                  }
                  if (strlen($lastcomment) > 0 or strlen($keywords) > 0) {
                    $this->output.=
                      "*<div class=\"bz_comment\">";
                    if ($keywords) {
                      $this->output.="<span class=\"bz_keyword\">$keywords</span>";
                    }
                    $this->output.=
                      "<span class=\"bug_id\">".
                      $line["id"].
                      "</span>\n".
                      $this->makeWikiSafe($lastcomment).
                      "</div>";
                  }
                }
                #
                # Render quick flag
                #
                if ($this->query->isRequired('quickflag')) {
                  if ($line["flagdate"]) {
                    $this->output.="<span class=\"flag\" title=\"Flag : ".
                      $line["flagdate"]."\">?</span>";
                  }
                }
              }         
              $this->output.="</td>";
            }
          }
          $this->output .= "</tr>";
        }
      }
    } 
      if ($this->query->get('detailsrow')) {
        $this->renderDetailsRow($detailsRowColumns,$line,
          $this->query->get('detailsrowprepend'));
      }
      #
      # We have LEFT JOIN so we need to ignore repeats
      #
      if ($this->query->get('blocks') && 
          ($line["blocks"] != $currentBlocksId)) {
        $currentBlocksId=$line["blocks"];     
        $this->renderDetailsRow(
        $this->query->blocksRowColumns,$line,"&rArr; ");
      }
      #
      # We have LEFT JOIN so we need to ignore repeats
      #
      if ($this->query->get('depends') && 
          ($line["depends"] != $currentDependsId)) {
        $currentDependsId=$line["depends"];     
        $this->renderDetailsRow(
        $this->query->dependsRowColumns,$line,"&lArr; ");
      }
    }

    # Total counter
    if ($this->query->get('total')) {
      $this->renderTotal($arrayOfTotals);
    }
      
    #
    # Display bar
    #
    if ($this->query->get('bar')) {
      $this->renderBar($barArray);
      $this->output.="<tr class=\"bz_bar_total\"><td colspan=\""
        .($this->numberOfMainRowColumns-1).
        "\">total</td><td class=\"total\">$nRows</td></tr>";      
    }
    
      # Table end
    if ($this->query->get('format') != "inline") {
        $this->output .= "</table>";
    }

    return $this->output;
  }
  
  #
  # Render a details row
  #
  private function renderDetailsRow($array,$line,$prepend) {
    $details="";
    $extra="";
    $title="";
    foreach (array_keys($array) as $column) {
      $dbColumn=$this->query->mapField($column);
      if ($line[$column]) {
        $valueTitle=$this->query->getValueTitle($line,$column);
        if ($valueTitle) {
          $this->context->debug &&
            $this->context
              ->debug("Setting title for $column to $valueTitle");
        }
        switch ($array[$column]) {
          case "title" :
            $title.=$this
              ->query->format($this->query->getDBValue($line,$dbColumn),
                $column,$valueTitle)." ";
            break;
          case "extra" :
            $extra.=$this
              ->query->format($this->query->getDBValue($line,$dbColumn),
                $column,$valueTitle)." ";
            break;
          default :
            $details.=$this
              ->query->format($this->query->getDBValue($line,$dbColumn),
                $column,$valueTitle)." ";
        }
      }
    } 
    if (trim($details)) {
      #
      # Use the trick in http://meta.wikimedia.org/wiki/Help:Sorting to allow the table
      # to be sorted
      #
      $this->output.="<tr class=\"bz_details\">";
      $i=0;
      foreach ($this->query->getColumns() as $column) {
        $dbColumn=$this->query->mapField($column);        
        if (!$this->query->get('detailsrow') 
            or !array_key_exists($dbColumn,
              $detailsRowColumns)) {
          $title=$this->query->getValueTitle($line,$dbColumn);
          $value=$this->query->
            format($this->query->getDBValue($line,$dbColumn),$column,$title);
          #
          # Putting ZZZ after the value ensures that it sorts after the row this is associated with
          $this->output.="<td><span style=\"display:none\">$value</span>";
          if ($i==0) {
            #
            # Don't close for the last column since we're going to append the details 
            #
            $estimatedWidth=30 + $this->numberOfMainRowColumns * 5;
            $this->output.="<div title=\"$title\" class=\"bz_details\" style=\"margin-left:2em;margin-right:-".$estimatedWidth."em;overflow:auto\">";
            $this->output.=$prepend.$details;
            if ($extra) {
              $this->output.="<span class=\"bz_extra\"> - $extra</span>";
            }
            $this->output.="</div>";
          }
          $this->output.="</td>";
          $i++;
        }
      }
      $this->output.="</tr>";
    }   
  }
  
  # Render the totals summary. 
  private function renderTotal($arrayOfTotals) {
    if ($this->query->get('format') != "inline") $this->output .= "<tr class=\"bz_total\">";
    $first=true;
    foreach ($this->query->getColumns() as $column) {
      if ($this->query->get('format') != "inline") {
        $this->output.="<td>";
      } 
      if (array_key_exists($column,$arrayOfTotals)) {
        if ($this->query->get('format') == "inline") {
          if ($first) {
            $first=false;
          } else {
            $this->output.=", ";
          }
        }
        $this->context->debug &&
          $this->context->debug("Writing $column total = ".
          $arrayOfTotals[$column]); 
        $this->output.=$arrayOfTotals[$column];
      } else {
        if ($this->query->get('format') != "inline") $this->output.="&nbsp;";
      }
      if ($this->query->get('format') != "inline") $this->output.="</td>";
    }
    if ($this->query->get('format') != "inline") $this->output.="</tr>";   
  }

  private function renderBar($barArray) {
    $total=0;
    $nonZeroKeyCount=0;
    $arrayKeys=array_keys($barArray);     
    foreach ($arrayKeys as $key) {
      $total+=$barArray[$key];
      $barArray[$key]>0 && $nonZeroKeyCount++;
    }
    $classColourCode="few";     
    $keyCount=count($arrayKeys);
    if ($keyCount > 6) {
      $classColourCode="some";
    } else if ($keyCount > 10) {
      $classColourCode="many";          
    }
    $this->output.="<tr class=\"bz_bar $classColourCode\"><td colspan=\""
      .($this->numberOfMainRowColumns-1).
      "\">";
    $i=0;
    $iNonZero=0;
    sort($arrayKeys);
    $rowTotal=0;
    $rowWidthTotal=0;
    foreach ($arrayKeys as $key) {
      $count=$barArray[$key];
      $i++;
      if ($count > 0) {
        $iNonZero++;
        $widthString="";
        $rowTotal+=$count;
        if ($iNonZero==$nonZeroKeyCount) {
          $class="C$i last";
          # Make sure we don't get caught by rounding error
          $width=100-$rowWidthTotal;  
        } else {
          $width=number_format((100*$count)/$total,0);
          $rowWidthTotal+=$width;
          $class="C$i notlast";
        }
        $content;
        if (strlen($key) > ($width /2) ) {
          $content="&nbsp;";
        } else {
          $keyAndCount="$key ($count)";
          if (strlen($keyAndCount) < ($width /2) ) {
            $content=$keyAndCount;
          } else {
            $content=$key;        
          }
        }
        $widthString=" style=\"width:".$width."%\"";            
        $this->output.="<div title=\"$key = $count";
        if ($this->context->debug) {
          $this->output.="class=$class / width=$width /  key count=$keyCount / nz key count=$nonZeroKeyCount / i=$i / inz=$iNonZero";
        }
        $this->output.=
          "\"$widthString class=\"$class\"><div>".
          "$content</div></div>";
      }
    } 
    $this->output.="</td><td class=\"total\">$rowTotal</td></tr>";
  }

  /**
   * Render output for no results
   */ 
  public function renderNoResultsHTML() {
    if ($this->query->get('noresultsmessage')) {
      return "<div class=\"bz_noresults\">".
        $this->query->get('noresultsmessage')."</div>";
    } else {
      return "";
    }
  }
  
  
  #
  # Initialise the bar array
  #
  private function getBarArray() {
    $barArray=array();
    if (array_key_exists($this->query->get('bar'),
        $this->query->fieldValues)) {
      foreach (split(",",
          $this->query->
            fieldValues[$this->query->get('bar')]) as $key) {
        $barArray[$key]=0;
      }
      $this->context->debug &&
        $this->context->debug("Initialised bar keys (".
          $this->query->get('bar')."): ".
          join(",",array_keys($barArray)));
    }
    return $barArray;
  }
  
  #
  # Make markup safe for inclusion in 
  #
  # (1) Append a "." at the end of wiki style headings so that the markup
  # does not get interpreted as wiki headings and get included in the TOC
  # See issue [#22]
  #
  private function makeWikiSafe($s) {
    $s=htmlspecialchars($s);
    if (strpos($s,"=") > -1) {
      $s=preg_replace('/^([=]+)[[:space:]]*$/m','<br/>$1<br/>',$s);
      # Bit cheating here because I don't make sure the closing equals match
      # the starting equals, but it's good enough
      $s=preg_replace('/^([=]+)([^=]*)=+[[:space:]]*$/m','$1$2$1.<br/>',$s);
    }
    return $s;
  }
}
?>
